go/vt/vterrors/main/main.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"log"
+	"os"
+	"strings"
+	"text/template"
+
+	"vitess.io/vitess/go/mysql"
+
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+const (
+	tmpl = `## Errors
+
+| ID | Description | Error | MySQL Error Code | SQL State |
+| --- | --- | --- | --- | --- |
+{{- range $err := . }}
+{{- $data := (call $err) }}
+| {{ $data.ID }} | {{ $data.Description }} | {{ FormatError $data.Err }} | {{ ConvertStateToMySQLErrorCode $data.State }} | {{ ConvertStateToMySQLState $data.State }} |
+{{- end }}
+`
+)
+
+// This program reads the errors located in the `vitess.io/vitess/go/vt/vterrors` package
+// and prints on the standard output a table, in Markdown format, that lists all the
+// errors with their code, description, error content, mysql error code and the SQL state.
+func main() {
```

Comment:
It's reasonably common for code generation use cases like this, so no problem with having an internal main.

A fairly common pattern I've seen is to add a build tag to omit it from the regular build. This still allows you to run the program via go run though:
```
matt@matt-3:~/tmp$ cat main.go 
//go:build ignore

package main

import "fmt"

func main() {
        fmt.Println("generated code!")
}
matt@matt-3:~/tmp$ go build
package mdlayher.com/tmp: build constraints exclude all Go files in /home/matt/tmp
1 matt@matt-3:~/tmp$ go run main.go
generated code!
```

One other suggestion would be to rename the folder this in from `main` to maybe something like `vterrorsgen` so that if a binary were to get built, it wouldn't be named `main`.

Reply:
> One other suggestion would be to rename the folder this in from main to maybe something like vterrorsgen so that if a binary were to get built, it wouldn't be named main.

That's a good idea! Thank you

> A pattern I've seen is to add a build tag to omit it from the regular build.

I don't think it's an issue if this binary is built along with all the other binaries. All our generation tools are built in the `bin/` directory already. This directive is good to know though, thank you 🥳

Reply:
Renamed the package in 36ebce6765

Reply:
Should also do that for `asthelpergen`, I noticed that it has a main directory.